src/vm/eventpipe.cpp

```diff
@@ -359,6 +361,7 @@ void EventPipe::DisableInternal(EventPipeSessionID id, EventPipeProviderCallback
             s_config.Disable(*pSession, pEventPipeProviderCallbackDataQueue);
         }
         pEventPipeThread->SetAsRundownThread(nullptr);
+        pSession->WriteAllBuffersToFile(); // Flush the rundown events to the stream/file
```

Comment:
Instead of suspending, then resuming, then suspending again I'd suggest just doing one suspend at line 371, followed by a call to WriteAllBuffers(). I'd imagine when your change, @cshung's change, and @sywhang's change all get composited together this code would look something like:

````
s_config.Disable(*pSession, …)

pSession->Disable(); // there is no longer a call to EPBM::SuspendWriteEvent in here
                     // @cshung change to set bit indicating session moving from active -> zombie occurs withinin or around Disable()
if(IsRundownRequested())
{
    pSession->WriteAllBuffersToFile() // this doesn't give a hard guarantee that all events have been drained to the file, but we didn't need one. We just want to decrease the odds of dropping rundown events by clearing space in the buffers. If rundown events get dropped anyways then the user is recommended to use a larger buffer size.

   pSession->EnableRundown();
   EventPipeThread *const pEventPipeThread = EventPipeThread::GetOrCreate();
    if (pEventPipeThread != nullptr)
    {
        pEventPipeThread->SetAsRundownThread(pSession);
        {
            s_config.Enable(*pSession, pEventPipeProviderCallbackDataQueue);
            {
                pSession->ExecuteRundown();
            }
            s_config.Disable(*pSession, pEventPipeProviderCallbackDataQueue);
        }
        pEventPipeThread->SetAsRundownThread(nullptr);
   }
   else
   {
       _ASSERTE(!"Failed to get or create the EventPipeThread for rundown events.");
       // return // this return should go away, we should finish cleaning up the session even if we couldn't do rundown
   }
}

pSession->SuspendWriteEvent() // assume this calls EPBM::SuspendWriteEvent()
/* this is the hard guarantee that all asynchronous writing is done. Theoretically a write could have started before Disable() was called, executed past all the checks if events were enabled or if the session was in a zombie state, but then switched out before actually writing the data into the event buffer and stayed that way through the entirety of rundown. SuspendWriteEvent will block until all such writes have either finished or aborted */

pSession->WriteAllBuffersToFile();
--s_numberOfSessions;
_ASSERTE(s_pSessions[pSession->GetIndex()] == pSession);
s_pSessions[pSession->GetIndex()].Store(nullptr);
pSession->WriteSequencePointUnbuffered();
delete pSession;
````

Reply:
In this new model, wouldn't changing
```
pSession->SuspendWriteEvent() // assume this calls EPBM::SuspendWriteEvent()
/* this is the hard guarantee that all asynchronous writing is done. Theoretically a write could have started before Disable() was called, executed past all the checks if events were enabled or if the session was in a zombie state, but then switched out before actually writing the data into the event buffer and stayed that way through the entirety of rundown. SuspendWriteEvent will block until all such writes have either finished or aborted */

pSession->WriteAllBuffersToFile();
--s_numberOfSessions;
_ASSERTE(s_pSessions[pSession->GetIndex()] == pSession);
s_pSessions[pSession->GetIndex()].Store(nullptr);
pSession->WriteSequencePointUnbuffered();
delete pSession;
```
to
```
_ASSERTE(s_pSessions[pSession->GetIndex()] == pSession);
s_pSessions[pSession->GetIndex()].Store(nullptr);
pSession->SuspendWriteEvent() // assume this calls EPBM::SuspendWriteEvent()
/* this is the hard guarantee that all asynchronous writing is done. Theoretically, a write could have started before Disable() was called, executed past all the checks if events were enabled or if the session was in a zombie state, but then switched out before actually writing the data into the event buffer and stayed that way through the entirety of rundown. SuspendWriteEvent will block until all such writes have either finished or aborted */

pSession->WriteAllBuffersToFile();
--s_numberOfSessions;
pSession->WriteSequencePointUnbuffered();
delete pSession;

```
solves the starvation problem and we don't need the zombie state at all?

Reply:
It works when we hold the EventPipe lock the whole time, but things go south if we try to break the lock. In particular assume that we don't want to be holding the lock across the call to pSession->WriteAllBuffersToFile() which could block for an arbitrarily long time if the listener isn't reading.
1) Since we aren't holding the lock a call to EventPipe::Enable() could occur.
2) The slot is free, so enable reallocates a new session in there
3) Threads attempt to write events to the new session, but inside EventPipeBufferManager::WriteEvent the call to thread.GetSessionState() returns the session state for the session we are in the midst of disabling. You can imagine things go downhill from here : ) Events get written to the wrong session and we probably get AVs when delete pSession eventually runs.

My original version actually has the same issue for pSession->WriteSequencePointUnbuffered() which should probably be moved before the s_pSessions[I] = nullptr. But after that swap we could safely put the EventPipe lock around just the null store and delete with no risk of blocking under the lock or re-using the slot prematurely.

FWIW I'm sure there are other ways to solve session state mismatch without having the zombie bitmask, although I'm not sure they'd be any simpler overall. There is also a second benefit of the zombie bitmask, it prevents non-rundown threads writing events into the session if the event provider/keyword is in the set of rundown provider/keywords. The rundown event provider/keywords aren't exclusively used by the implementation of rundown unfortunately. This one could also be solved other ways, its just a matter of what is the simplest solution we can find that covers it all. If we find a simpler one I'd be glad to use it.

Reply:
I noticed that `EventPipeSession::Disable` already called `WriteAllBuffersToFile` so I removed the extra call I was making and noted in a comment that the write is a side-effect of disable.